nexus/src/app/oximeter.rs

```diff
@@ -113,55 +112,42 @@ impl super::Nexus {
-            let db_info = db::model::ProducerEndpoint::new(&producer_info, id);
+        let collector_id = self
+            .db_datastore
+            .producer_endpoint_create(opctx, &producer_info)
+            .await?;
 
-            // We chose the collector in `self.next_collector` above; if we get
-            // an "Oximeter not found" error when we try to create a producer
-            // assigned to that collector, we've lost an extremely rare race
-            // where the collector we chose was deleted in between when we chose
-            // it and when we tried to assign it. If we hit this, we should just
-            // pick another collector and try again.
-            //
-            // To safeguard against some other bug forcing us into an infinite
-            // loop here, we'll only retry once. Losing this particular race
-            // once is exceedingly unlikely; losing it twice probably means
-            // something else is wrong, so we'll just return the error.
-            match self
-                .db_datastore
-                .producer_endpoint_create(opctx, &db_info)
-                .await
-            {
-                Ok(()) => (), // fallthrough
-                Err(Error::ObjectNotFound {
-                    type_name: ResourceType::Oximeter,
-                    lookup_type: LookupType::ById(bad_id),
-                }) if id == bad_id && attempt == 0 => {
-                    // We lost the race on our first try; try again.
-                    continue;
-                }
-                // Any other error or we lost the race twice; fail.
-                Err(err) => return Err(err),
-            }
+        // It's possible (albeit quite unlikely) that the collector we chose in
+        // the query above was expunged between then and when we execute this
+        // query to look up its information. In such a case, we've now assigned
+        // this producer to an expunged collector, so failing is fine - we'll
+        // rely on reconfigurator to come back in and reassign this producer
+        // once a new oximeter is available (just like it would have to do for
+        // for any producers that were assigned to this collector already).
+        let collector_info =
+            self.db_datastore.oximeter_lookup(opctx, &collector_id).await?;
 
-            collector
-                .producers_post(
-                    &oximeter_client::types::ProducerEndpoint::from(
-                        &producer_info,
-                    ),
-                )
-                .await
-                .map_err(Error::from)?;
-            info!(
-                self.log,
-                "assigned collector to new producer";
-                "producer_id" => ?producer_info.id,
-                "collector_id" => ?id,
-            );
-            return Ok(());
-        }
-        unreachable!("for loop always returns after at most two iterations")
+        let address = SocketAddr::from((
```

Comment:
Not sure it's worth the headache, but you could return this directly from the query I _think_. In the subquery that chooses the collector ID, we could return the address as well, and then use that in the `RETURNING` clause.

I could be wrong about that, because the `RETURNING` clause applies to the inserted values in the `producer_endpoint` table, which doesn't have that address as a column.

Reply:
We can't do it in that subquery, but it's easy enough to add another clause after this that uses our returned `oximeter_id` to select from `oximeter` to get the other columns.

After doing that, I realized I wasn't testing for the somewhat unusual case I ran into while testing #6748 - what does this query do when the oximeter is expunged? I added a test for that and realized that it's inconsistent:

* If there is only one oximeter and it is expunged, this query fails. That matches the behavior I saw on #6748 and would cause producers to be pruned if we don't replace the oximeter fast enough. (Which, as we discussed offline, is fine.)
* If we're currently assigned to an expunged oximeter, but there is _also_ a non-expunged oximeter, this query actually succeeds, but in a surprising (and probably wrong) way: it updates `time_modified` but does _not_ change `oximeter_id`, so it refreshes the producer without reassigning it.

I started trying to "fix" the second case to make it fail like the first case does. But in doing so, I realized it's more straightforward to just also reassign `oximeter_id` if we're in the second case (where we're assigned to an expunged oximeter, but there is at least one non-expunged oximeter available). As a side benefit, I think that makes the blueprint execution RPW's "reassign all producers" bit purely a performance optimization - with the reworked query, as the producers reregister themselves, Nexus will also reassign them.

This wasn't a trivial rework, so the query is probably worth a second look, if you don't mind.